pylint/checkers/variables.py

```diff
@@ -617,9 +639,27 @@ def get_next_to_consume(self, node):
             found_nodes = [
                 n
                 for n in found_nodes
-                if not isinstance(n.statement(), nodes.ExceptHandler)
-                or n.statement().parent_of(node)
+                if not isinstance(n.statement(future=True), nodes.ExceptHandler)
+                or n.statement(future=True).parent_of(node)
+            ]
+
+        # Filter out assignments in an Except clause that the node is not
+        # contained in, assuming they may fail
+        if found_nodes:
+            filtered_nodes = [
+                n
+                for n in found_nodes
+                if not (
+                    isinstance(n.statement(future=True).parent, nodes.ExceptHandler)
+                    and isinstance(
+                        n.statement(future=True).parent.parent, nodes.TryExcept
+                    )
+                )
+                or n.statement(future=True).parent.parent_of(node)
             ]
+            difference = [n for n in found_nodes if n not in filtered_nodes]
```

Comment:
This is quite inefficient, a list contains check for each node.
I would suggest to create a temporary set from `filtered_nodes` to do the check with.

Reply:
Something like this?
```diff
diff --git a/pylint/checkers/variables.py b/pylint/checkers/variables.py
index 6559d07f..0e5a0fd1 100644
--- a/pylint/checkers/variables.py
+++ b/pylint/checkers/variables.py
@@ -657,8 +657,8 @@ scope_type : {self._atomic.scope_type}
                 )
                 or n.statement(future=True).parent.parent_of(node)
             ]
-            difference = [n for n in found_nodes if n not in filtered_nodes]
-            self.consumed_uncertain[node.name] += difference
+            difference = set(found_nodes) - set(filtered_nodes)
+            self.consumed_uncertain[node.name] += list(difference)
             found_nodes = filtered_nodes
 
         return found_nodes
```

Reply:
The most performant solution would be to build two arrays while the condition above is being checked, but that didn't rhyme with the neighboring patterns, so I didn't pursue it initially. What do you think?

Reply:
I'm not sure if the order in difference is important. If so, it's lost with your suggestion. I was thinking more along these lines here. It's also not the optimal solution but should be good enough.
```py
            filtered_nodes_set = set(filtered_nodes)
            difference = [n for n in found_nodes if n not in filtered_nodes_set]
            self.consumed_uncertain[node.name] += difference
```